cli/internal/context/context.go

```diff
@@ -379,40 +373,50 @@ func (c *Context) parsePackageJSON(repoRoot turbopath.AbsoluteSystemPath, pkgJSO
 	return nil
 }
 
-func (c *Context) resolveDepGraph(wg *errgroup.Group, workspace *fs.PackageJSON, unresolvedDirectDeps map[string]string, resolvedDepsSet mapset.Set, seen mapset.Set, pkg *fs.PackageJSON) {
-	if c.Lockfile == (lockfile.Lockfile)(nil) {
-		return
+// TransitiveClosure the set of all lockfile keys that pkg depends on
+func TransitiveClosure(pkg *fs.PackageJSON, lockFile lockfile.Lockfile) (mapset.Set, error) {
+	if lockfile.IsNil(lockFile) {
+		return nil, fmt.Errorf("No lockfile available to do analysis on")
+	}
+
+	resolvedPkgs := mapset.NewSet()
+	lockfileEg := &errgroup.Group{}
+
+	transitiveClosureHelper(lockfileEg, pkg, lockFile, pkg.UnresolvedExternalDeps, resolvedPkgs)
+
+	if err := lockfileEg.Wait(); err != nil {
+		return nil, err
 	}
+
+	return resolvedPkgs, nil
+}
+
+func transitiveClosureHelper(wg *errgroup.Group, pkg *fs.PackageJSON, lockfile lockfile.Lockfile, unresolvedDirectDeps map[string]string, resolvedDeps mapset.Set) {
 	for directDepName, unresolvedVersion := range unresolvedDirectDeps {
 		directDepName := directDepName
 		unresolvedVersion := unresolvedVersion
 		wg.Go(func() error {
 
-			lockfilePkg, err := c.Lockfile.ResolvePackage(workspace.Dir.ToUnixPath(), directDepName, unresolvedVersion)
+			lockfilePkg, err := lockfile.ResolvePackage(pkg.Dir.ToUnixPath(), directDepName, unresolvedVersion)
 
 			if err != nil {
 				return err
 			}
 
-			if !lockfilePkg.Found || seen.Contains(lockfilePkg.Key) {
+			if !lockfilePkg.Found || resolvedDeps.Contains(lockfilePkg) {
 				return nil
 			}
 
-			seen.Add(lockfilePkg.Key)
-
-			pkg.Mu.Lock()
-			pkg.TransitiveDeps = append(pkg.TransitiveDeps, lockfilePkg.Key)
-			pkg.Mu.Unlock()
-			resolvedDepsSet.Add(fmt.Sprintf("%s@%s", lockfilePkg.Key, lockfilePkg.Version))
+			resolvedDeps.Add(lockfilePkg)
```

Comment:
Yeah, we needed the lock because each goroutine that was spun up via the working group could try to append `TransitiveDeps` and go slices aren't thread safe. `mapset.Set` is threadsafe so we don't need to lock anything up